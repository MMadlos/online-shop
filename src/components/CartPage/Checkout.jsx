import Chip from "../Elements/Chip"
import shippingIMG from "../../assets/free-shipping-tiny.png"
import Button from "../Elements/Button"

function Checkout({ totalToPay, totalItems }) {
	return (
		<div className="checkout-container">
			<img
				src={shippingIMG}
				alt=""
				className="shipping-image"
			/>
			<div className="cost-list">
				<div className="subtotal">
					<p>Number of items</p>
					<p>{totalItems}</p>
				</div>
				<div className="subtotal">
					<p>Subtotal</p>
					<p>$ {totalToPay}</p>
				</div>
				<div className="divider"></div>
				<div className="shipping-cost">
					<p>Shipping cost</p>
					<Chip
						text="FREE"
						type="accent"
					/>
				</div>
				<div className="total-amount">
					<p>TOTAL</p>
					<p>$ {totalToPay}</p>
				</div>
				<p className="taxes">(Taxes included)</p>
			</div>
			<Button
				text="Checkout"
				btnID="checkout"
			/>
		</div>
	)
}

export default Checkout
