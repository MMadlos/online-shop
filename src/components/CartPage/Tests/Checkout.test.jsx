import { describe, it, expect, vi } from "vitest"
import { render, screen } from "@testing-library/react"
import { BrowserRouter } from "react-router-dom"

import Checkout from "../Checkout"
import userEvent from "@testing-library/user-event"

const totalToPay = 2
const totalItems = 1

describe("Checkout", () => {
	it.skip("renders component", () => {
		const { container } = render(
			<Checkout
				totalToPay={totalToPay}
				totalItems={totalItems}
			/>
		)
		expect(container).toMatchSnapshot()
	})
	it("renders subtotal, total and items number", () => {
		render(
			<Checkout
				totalToPay={totalToPay}
				totalItems={totalItems}
			/>
		)

		const totalItemsEl = screen.getByTestId("total-items")
		const subtotalAmountEl = screen.getByTestId("subtotal-amount")
		const totalAmountEl = screen.getByTestId("total-amount")

		expect(totalItemsEl).toBeInTheDocument()
		expect(totalItemsEl).toHaveTextContent(totalItems)

		expect(subtotalAmountEl).toBeInTheDocument()
		expect(subtotalAmountEl).toHaveTextContent(totalToPay)

		expect(totalAmountEl).toBeInTheDocument()
		expect(totalAmountEl).toHaveTextContent(totalToPay)
	})

	it("renders image with the corresponding url", () => {
		render(
			<Checkout
				totalToPay={totalToPay}
				totalItems={totalItems}
			/>
		)

		const img = screen.getByRole("img")
		expect(img).toHaveAttribute("src", expect.not.stringContaining(" "))
	})

	it("renders button with id = checkout", () => {
		render(
			<Checkout
				totalToPay={totalToPay}
				totalItems={totalItems}
			/>
		)

		const checkoutBtn = screen.getByRole("button")
		expect(checkoutBtn).toBeInTheDocument()
		expect(checkoutBtn).toHaveAttribute("id", "checkout")
		expect(checkoutBtn).toHaveTextContent(/checkout/i)
	})

	it("opens CheckoutPage if button is clicked", async () => {
		vi.mock("../CheckoutPage", () => ({
			default: () => {
				return <div data-testid="checkout-page">Checkout Page Content</div>
			},
		}))

		const user = userEvent.setup()

		render(
			<BrowserRouter>
				<Checkout
					totalToPay={totalToPay}
					totalItems={totalItems}
				/>
			</BrowserRouter>
		)

		const checkoutBtn = screen.getByRole("button", { name: /checkout/i })
		await user.click(checkoutBtn)
		screen.debug()

		const checkoutPage = screen.getByTestId("checkout-page")

		expect(checkoutPage).toBeInTheDocument()
		expect(checkoutPage).toHaveTextContent("Checkout Page")
	})

	it("doesn't open CheckoutPage if button is not clicked", () => {
		render(
			<BrowserRouter>
				<Checkout
					totalToPay={totalToPay}
					totalItems={totalItems}
				/>
			</BrowserRouter>
		)

		const checkoutPage = screen.queryByTestId("checkout-page")
		expect(checkoutPage).not.toBeInTheDocument()
	})
})
