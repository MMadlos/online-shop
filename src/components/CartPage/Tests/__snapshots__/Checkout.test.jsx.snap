// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Checkout > renders component 1`] = `
<div>
  <div
    class="checkout-container"
  >
    <img
      alt=""
      class="shipping-image"
      src="/src/assets/free-shipping-tiny.png"
    />
    <div
      class="cost-list"
    >
      <div
        class="subtotal"
      >
        <p>
          Number of items
        </p>
        <p>
          1
        </p>
      </div>
      <div
        class="subtotal"
      >
        <p>
          Subtotal
        </p>
        <p>
          $ 
          2
        </p>
      </div>
      <div
        class="divider"
      />
      <div
        class="shipping-cost"
      >
        <p>
          Shipping cost
        </p>
        <div
          class="chip accent"
        >
          <p>
            FREE
          </p>
        </div>
      </div>
      <div
        class="total-amount"
      >
        <p>
          TOTAL
        </p>
        <p>
          $ 
          2
        </p>
      </div>
      <p
        class="taxes"
      >
        (Taxes included)
      </p>
    </div>
    <button
      id="checkout"
    >
      Checkout
    </button>
  </div>
</div>
`;
